### PR TITLE
Search for name

### PR DESCRIPTION
**Pull request by pellaea**
Changes behavior of plain old observation search for a name.

Type a name like "Xanthoria elegans" into the search bar, then search for observations.
It used to return only observations of Xanthoria elegans, not its synonym "Rusavskia elegans".
I am arguing that the vast majority of users will expect it to include synonyms by default.

Also, I think that the majority of users, at least new users, will expect a search for "Amanita" to show all observations of Amanita, not just those too poor to identify to species.  Therefore this PR now makes the default behavior of all searches with the names parameter supplied to be include_subtaxa:yes.

This PR changes it so that the observation pattern search recognizes that the pattern the user entered is a name (checks for that name in the database under text_name and search_name), then moves the pattern parameter into the names parameter.  And if the names parameter is present, it supplies defaults for both include_synonyms and include_subtaxa.

### CODE REVIEW

**Comment by JoeCohen** [comment]
Jason: Thanks so much for doing this. 
It's how Observation searches should function; it's how other mycological sites (e.g. MyCoPortal), function; and is a long-requested feature. See, e.g., https://www.pivotaltracker.com/story/show/164957447; https://www.pivotaltracker.com/story/show/25304399.

File: app/classes/pattern_search/observation.rb [hunk 1]
```diff
@@ -66,16 +66,43 @@ def model
...
-          !args[:include_all_name_proposals].nil?)
-        self.flavor = :all
-        args[:names] = args[:pattern]
-        args.delete(:pattern)
-      end
+    def any_taxa_modifiers_present?
+      !args[:include_subtaxa].nil? ||
+        !args[:include_synonyms].nil? ||
+        !args[:include_all_name_proposals].nil? ||
+        !args[:exclude_consensus].nil?
```
Comment (JoeCohen) [inline app/classes/pattern_search/observation.rb:105]: Could any of these be empty, but not nil? 
If so, then 
```suggestion
      args[:include_subtaxa].present? ||
      args[:include_synonyms].present? ||
      args[:include_all_name_proposals].present?
      args[:exclude_consensus].present?
```
If not, will `args[x]` work in place of `!args[x].nil?`?
```suggestion
      args[:include_subtaxa] ||
      args[:include_synonyms] ||
      args[:include_all_name_proposals] ||
      args[:exclude_consensus]
```

Comment (pellaea) [reply app/classes/pattern_search/observation.rb:105]: As I understand it args[:boolean] can have three states: nil, true or false.  Therefore neither of your proposals would work, I'm afraid.  Actually, I already had tried your first proposal!  (great minds obviously think alike)  I suppose it *could* be re-written as !(a.nil? && b.nil? && c.nil?).  But that doesn't really feel any better to me.

Comment (JoeCohen) [reply app/classes/pattern_search/observation.rb:105]: Thanks for the explanation. I hadn't realized what you were trying to accomplish.

**Comment by pellaea** [comment]
Ugh, the failing tests all had to do with query titles.  That is definitely a larger problem than I have time to solve completely.  (How do we get observation queries to describe themselves in English when there are so many possible permutations of parameters??)  But I did take a stab at it, trying at least to preserve the titles for the flavor=all versions of the old flavors like by_user, of_name, at_location, etc.  Almost certainly not perfect, but at least all the tests pass again!

This is definitely something that needs to happen eventually if we ever want to convert all those old "flavorful" queries into "unflavored" (i.e. flavor = :all) queries using filter parameters instead.  (Which would be a useful way to make various queries and indexes more transparent and RESTful and persistent.  "Observations By User" then becomes a generic observation query with "user=xxx" filter parameter, and we can automatically fill in the search bar with "user:xxx".  Then it will be very easy for users to modify the query in nontrivial and useful ways by simply adding terms to the search bar.  And so on and on.  You get the point, I'm sure.  Useful direction to move, I think, just going to be a long journey...

**Comment by JoeCohen** [comment]
There are some issues. Here are a couple. (Given these issues I didn't look at query titles at all.)
#### Search for Observations of `Gomphus floccosus`:
(I.e., type `Gomphus floccosus` in the pattern search bar, Search for Observations.)
>No matching observations found.
Other Links:
Observations of [_Gomphus floccosus_ (Schwein.) Singer (398)](http://localhost:3000/observer/observation_search?pattern=Gomphus+floccosus)

Note:  Clicking on the link in "Other Links" lands me on the right page.
_G. floccosus_ is a deprecated Name for _Turbinellus floccosus_

#### Search for Observations of `Turbinellus floccosus` 
sends me to a single Obs (282311) of a different taxon: _Turbinellus **kauffmanii**_.  
(It should return 398 Observations.)

**Comment by JoeCohen** [comment]
I cannot replicate the 2nd problem. Weird. (Maybe related to my starting state?)

**Comment by JoeCohen** [comment]
As to the first example, the log follows. 
I wonder about 
```sql
LEFT OUTER JOIN `locations` ON observations.location_id = locations.id
```
Also whether it has to look for all those regex's when if there's a single exact `text_name` match.
```sql
Started POST "/observer/pattern_search" for 10.0.2.2 at 2020-12-10 23:37:24 +0000
   (0.6ms)  SET  @@SESSION.sql_mode = 'TRADITIONAL',  @@SESSION.sql_auto_is_null = 0, @@SESSION.wait_timeout = 2147483
   (32.9ms)  SELECT `schema_migrations`.`version` FROM `schema_migrations` ORDER BY `schema_migrations`.`version` ASC
Creating scope :thumbnails. Overwriting existing method User.thumbnails.
Creating scope :smalls. Overwriting existing method User.smalls.
Processing by ObserverController#pattern_search as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"yYvgozw/5uRC+/boULwlHw2RRzflRYuBkGwhxyeSOWHmtYlQdf4uBcIJBdUp89JpikZr1o2dNHINKQ3kUA3RFQ==", "search"=>{"pattern"=>"gompus floccosus", "type"=>"observation"}, "commit"=>"Search"}
  User Load (0.6ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 4468 LIMIT 1
  User Load (0.8ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 4468 LIMIT 1
user=4468 robot=N ip=10.0.2.2
   (0.2ms)  BEGIN
  User Load (10.2ms)  SELECT  `users`.* FROM `users` WHERE `users`.`login` = 'Joe Cohen' LIMIT 1
  User Update (1.1ms)  UPDATE `users` SET `last_activity` = '2020-12-10 23:37:27', `updated_at` = '2020-12-10 23:37:27' WHERE `users`.`id` = 4468
   (39.0ms)  COMMIT
[I18n] loading locale: en from @user
  Language Load (0.4ms)  SELECT  `languages`.* FROM `languages` WHERE `languages`.`locale` = 'en' LIMIT 1
[I18n] Locale set to en
   (18.9ms)  
      SELECT locale, tag, text
      FROM translation_strings t, languages l
      WHERE t.language_id = l.id
        AND t.updated_at >= '2020-12-10 23:36:27.931387'
    
  Language Load (0.6ms)  SELECT  `languages`.* FROM `languages` WHERE `languages`.`locale` = 'en' LIMIT 1
  UserGroup Load (89.9ms)  SELECT `user_groups`.* FROM `user_groups` INNER JOIN `user_groups_users` ON `user_groups`.`id` = `user_groups_users`.`user_group_id` WHERE `user_groups_users`.`user_id` = 4468
Redirected to http://localhost:3000/observer/observation_search?pattern=gompus+floccosus
TIME: 0.694974329 302 observer pattern_search user 10.0.2.2	http://localhost:3000/observer/pattern_search	Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36
Completed 302 Found in 697ms (ActiveRecord: 216.3ms)


10.0.2.2 - - [10/Dec/2020:23:37:24 UTC] "POST /observer/pattern_search HTTP/1.1" 302 140
http://localhost:3000/observer/observation_search?pattern=gompus+floccosus -> /observer/pattern_search
Started GET "/observer/observation_search?pattern=gompus+floccosus" for 10.0.2.2 at 2020-12-10 23:37:28 +0000
Processing by ObserverController#observation_search as HTML
  Parameters: {"pattern"=>"gompus floccosus"}
  User Load (0.7ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 4468 LIMIT 1
  User Load (0.9ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 4468 LIMIT 1
user=4468 robot=N ip=10.0.2.2
   (0.2ms)  BEGIN
  User Load (8.0ms)  SELECT  `users`.* FROM `users` WHERE `users`.`login` = 'Joe Cohen' LIMIT 1
  User Update (28.1ms)  UPDATE `users` SET `last_activity` = '2020-12-10 23:37:28', `updated_at` = '2020-12-10 23:37:28' WHERE `users`.`id` = 4468
   (1.8ms)  COMMIT
[I18n] loading locale: en from @user
  Language Load (0.6ms)  SELECT  `languages`.* FROM `languages` WHERE `languages`.`locale` = 'en' LIMIT 1
[I18n] Locale set to en
   (15.1ms)  
      SELECT locale, tag, text
      FROM translation_strings t, languages l
      WHERE t.language_id = l.id
        AND t.updated_at >= '2020-12-10 23:37:27.964123'
    
  Language Load (0.5ms)  SELECT  `languages`.* FROM `languages` WHERE `languages`.`locale` = 'en' LIMIT 1
  UserGroup Load (92.9ms)  SELECT `user_groups`.* FROM `user_groups` INNER JOIN `user_groups_users` ON `user_groups`.`id` = `user_groups_users`.`user_group_id` WHERE `user_groups_users`.`user_id` = 4468
  Name Exists (104.7ms)  SELECT  1 AS one FROM `names` WHERE (text_name = 'gompus floccosus' OR search_name = 'gompus floccosus') LIMIT 1
  QueryRecord Load (0.7ms)  SELECT  `query_records`.* FROM `query_records` WHERE `query_records`.`description` = 'flavor=:pattern_search;model=:Observation;pattern=$gompus floccosus' LIMIT 1
   (0.8ms)  
      DELETE FROM query_records
      WHERE
        access_count = 0 AND updated_at < DATE_SUB(NOW(), INTERVAL 6 HOUR) OR
        access_count > 0 AND updated_at < DATE_SUB(NOW(), INTERVAL 1 DAY)
    
   (2873.5ms)  
      SELECT count(*)
      FROM `observations` LEFT OUTER JOIN `locations` ON observations.location_id = locations.id JOIN `names` ON observations.name_id = names.id
      WHERE CONCAT(names.search_name,COALESCE(observations.notes,''),observations.where) LIKE '%gompus%' AND CONCAT(names.search_name,COALESCE(observations.notes,''),observations.where) LIKE '%floccosus%'
  ORDER BY observations.when DESC, observations.id DESC

QUERY starting: "\n      SELECT DISTINCT observations.id\n      FROM `observations` LEFT OUTER JOIN `locations` ON observations.location_id = locations.id JOIN `names` ON observations.name_id = names.id\n      WHERE CONCAT(names.search_name,COALESCE(observations.notes,''),observations.where) LIKE '%gompus%' AND CONCAT(names.search_name,COALESCE(observations.notes,''),observations.where) LIKE '%floccosus%'\n  ORDER BY observations.when DESC, observations.id DESC\n"
   (2791.4ms)  
      SELECT DISTINCT observations.id
      FROM `observations` LEFT OUTER JOIN `locations` ON observations.location_id = locations.id JOIN `names` ON observations.name_id = names.id
      WHERE CONCAT(names.search_name,COALESCE(observations.notes,''),observations.where) LIKE '%gompus%' AND CONCAT(names.search_name,COALESCE(observations.notes,''),observations.where) LIKE '%floccosus%'
  ORDER BY observations.when DESC, observations.id DESC
  LIMIT 0,120

QUERY finished: model=Observation, flavor=pattern_search, params={:pattern=>"gompus floccosus"}, time=2.792646259
  Rendering observer/list_observations.html.erb within layouts/application
  Name Load (170.4ms)  SELECT  `names`.* FROM `names` WHERE ((LENGTH(text_name) BETWEEN '5' AND '9') AND (text_name LIKE '%ompus' OR text_name LIKE 'G%mpus' OR text_name LIKE 'Go%pus' OR text_name LIKE 'Gom%us' OR text_name LIKE 'Gomp%s' OR text_name LIKE 'Gompu%') AND correct_spelling_id IS NULL) LIMIT 10
  Name Load (209.1ms)  SELECT  `names`.* FROM `names` WHERE ((LENGTH(text_name) BETWEEN '15' AND '19') AND (text_name LIKE '%omphus floccosus' OR text_name LIKE 'G%mphus floccosus' OR text_name LIKE 'Go%phus floccosus' OR text_name LIKE 'Gom%hus floccosus' OR text_name LIKE 'Gomp%us floccosus' OR text_name LIKE 'Gomph%s floccosus' OR text_name LIKE 'Gomphu% floccosus' OR text_name LIKE 'Gomphus %loccosus' OR text_name LIKE 'Gomphus f%occosus' OR text_name LIKE 'Gomphus fl%ccosus' OR text_name LIKE 'Gomphus flo%cosus' OR text_name LIKE 'Gomphus floc%osus' OR text_name LIKE 'Gomphus flocc%sus' OR text_name LIKE 'Gomphus flocco%us' OR text_name LIKE 'Gomphus floccos%s' OR text_name LIKE 'Gomphus floccosu%') AND correct_spelling_id IS NULL) LIMIT 10
  Name Load (236.4ms)  SELECT  `names`.* FROM `names` WHERE ((LENGTH(text_name) BETWEEN '17' AND '21') AND (text_name LIKE '%omphales floccosus' OR text_name LIKE 'G%mphales floccosus' OR text_name LIKE 'Go%phales floccosus' OR text_name LIKE 'Gom%hales floccosus' OR text_name LIKE 'Gomp%ales floccosus' OR text_name LIKE 'Gomph%les floccosus' OR text_name LIKE 'Gompha%es floccosus' OR text_name LIKE 'Gomphal%s floccosus' OR text_name LIKE 'Gomphale% floccosus' OR text_name LIKE 'Gomphales %loccosus' OR text_name LIKE 'Gomphales f%occosus' OR text_name LIKE 'Gomphales fl%ccosus' OR text_name LIKE 'Gomphales flo%cosus' OR text_name LIKE 'Gomphales floc%osus' OR text_name LIKE 'Gomphales flocc%sus' OR text_name LIKE 'Gomphales flocco%us' OR text_name LIKE 'Gomphales floccos%s' OR text_name LIKE 'Gomphales floccosu%') AND correct_spelling_id IS NULL) LIMIT 10
  Name Load (219.8ms)  SELECT  `names`.* FROM `names` WHERE ((LENGTH(text_name) BETWEEN '17' AND '21') AND (text_name LIKE '%mphales floccosus' OR text_name LIKE 'G%phales floccosus' OR text_name LIKE 'Go%hales floccosus' OR text_name LIKE 'Gom%ales floccosus' OR text_name LIKE 'Gomp%les floccosus' OR text_name LIKE 'Gomph%es floccosus' OR text_name LIKE 'Gompha%s floccosus' OR text_name LIKE 'Gomphal% floccosus' OR text_name LIKE 'Gomphales %occosus' OR text_name LIKE 'Gomphales f%ccosus' OR text_name LIKE 'Gomphales fl%cosus' OR text_name LIKE 'Gomphales flo%osus' OR text_name LIKE 'Gomphales floc%sus' OR text_name LIKE 'Gomphales flocc%us' OR text_name LIKE 'Gomphales flocco%s' OR text_name LIKE 'Gomphales floccos%') AND correct_spelling_id IS NULL) LIMIT 10
  Name Load (218.0ms)  SELECT  `names`.* FROM `names` WHERE ((LENGTH(text_name) BETWEEN '17' AND '21') AND (text_name LIKE '%phales floccosus' OR text_name LIKE 'G%hales floccosus' OR text_name LIKE 'Go%ales floccosus' OR text_name LIKE 'Gom%les floccosus' OR text_name LIKE 'Gomp%es floccosus' OR text_name LIKE 'Gomph%s floccosus' OR text_name LIKE 'Gompha% floccosus' OR text_name LIKE 'Gomphales %ccosus' OR text_name LIKE 'Gomphales f%cosus' OR text_name LIKE 'Gomphales fl%osus' OR text_name LIKE 'Gomphales flo%sus' OR text_name LIKE 'Gomphales floc%us' OR text_name LIKE 'Gomphales flocc%s' OR text_name LIKE 'Gomphales flocco%') AND correct_spelling_id IS NULL) LIMIT 10
  Name Load (209.4ms)  SELECT  `names`.* FROM `names` WHERE ((LENGTH(text_name) BETWEEN '15' AND '19') AND (text_name LIKE '%omphos floccosus' OR text_name LIKE 'G%mphos floccosus' OR text_name LIKE 'Go%phos floccosus' OR text_name LIKE 'Gom%hos floccosus' OR text_name LIKE 'Gomp%os floccosus' OR text_name LIKE 'Gomph%s floccosus' OR text_name LIKE 'Gompho% floccosus' OR text_name LIKE 'Gomphos %loccosus' OR text_name LIKE 'Gomphos f%occosus' OR text_name LIKE 'Gomphos fl%ccosus' OR text_name LIKE 'Gomphos flo%cosus' OR text_name LIKE 'Gomphos floc%osus' OR text_name LIKE 'Gomphos flocc%sus' OR text_name LIKE 'Gomphos flocco%us' OR text_name LIKE 'Gomphos floccos%s' OR text_name LIKE 'Gomphos floccosu%') AND correct_spelling_id IS NULL) LIMIT 10
  Name Exists (0.7ms)  SELECT  1 AS one FROM `names` WHERE (text_name = 'Gomphus floccosus' OR search_name = 'Gomphus floccosus') LIMIT 1
  QueryRecord Load (0.8ms)  SELECT  `query_records`.* FROM `query_records` WHERE `query_records`.`description` = 'flavor=:all;include_subtaxa=1;include_synonyms=1;model=:Observation;names=@$Gomphus floccosus' LIMIT 1
  Name Load (142.8ms)  SELECT `names`.* FROM `names` WHERE `names`.`text_name` = 'Gomphus floccosus'
   (40.7ms)  
            SELECT id, correct_spelling_id, synonym_id, text_name FROM names
            WHERE synonym_id IN (132)
          
   (128.2ms)  
          SELECT id, correct_spelling_id, synonym_id, text_name FROM names
          WHERE text_name REGEXP '^(Gomphus floccosus|Gomphus bonarii|Gomphus thiersii|Turbinellus floccosus|Cantharellus bonarii) '
        
   (39.4ms)  
            SELECT id, correct_spelling_id, synonym_id, text_name FROM names
            WHERE synonym_id IN (132,6466)
          
   (281.2ms)  
      SELECT count(*)
      FROM `observations` 
      WHERE observations.name_id IN (254,947,4489,5529,40065,44749,61193)
  ORDER BY observations.when DESC, observations.id DESC

  Rendered collection of shared/_matrix_box.html.erb [0 times] (0.0ms)
  Rendered observer/list_observations.html.erb within layouts/application (2376.4ms)
  TranslationString Load (26.1ms)  SELECT `translation_strings`.* FROM `translation_strings` WHERE `translation_strings`.`tag` = 'title_for_observation_search'
  Language Load (1.1ms)  SELECT  `languages`.* FROM `languages` WHERE `languages`.`official` = TRUE LIMIT 1
  TranslationString Load (10.0ms)  SELECT  `translation_strings`.* FROM `translation_strings` WHERE `translation_strings`.`tag` = 'app_banner_box' AND `translation_strings`.`language_id` = 2 ORDER BY `translation_strings`.`id` ASC LIMIT 1
  Language Load (0.6ms)  SELECT `languages`.* FROM `languages`
  Rendered shared/_nav_left.html.erb (20.0ms)
  Rendered shared/_search_bar.html.erb (141.3ms)
  Rendered shared/_title_and_tab_sets.html.erb (2.5ms)
  Language Load (0.7ms)  SELECT  `languages`.* FROM `languages` WHERE `languages`.`locale` = 'en' LIMIT 1
  Rendered shared/_translator_credit.html.erb (2.4ms)
TIME: 12.533050081 200 observer observation_search user 10.0.2.2	http://localhost:3000/observer/observation_search?pattern=gompus+floccosus	Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36
Completed 200 OK in 12534ms (Views: 2987.3ms | ActiveRecord: 7907.2ms)


10.0.2.2 - - [10/Dec/2020:23:37:28 UTC] "GET /observer/observation_search?pattern=gompus+floccosus HTTP/1.1" 200 13977
http://localhost:3000/observer/observation_search?pattern=gompus+floccosus -> /observer/observation_search?pattern=gompus+floccosus
Started GET "/assets/Hygrocybe.css" for 10.0.2.2 at 2020-12-10 23:37:41 +0000
Started GET "/assets/application.js" for 10.0.2.2 at 2020-12-10 23:37:41 +0000
   (1.7ms)  SET  @@SESSION.sql_mode = 'TRADITIONAL',  @@SESSION.sql_auto_is_null = 0, @@SESSION.wait_timeout = 2147483
10.0.2.2 - - [10/Dec/2020:23:37:41 UTC] "GET /assets/application.js HTTP/1.1" 304 0
http://localhost:3000/observer/observation_search?pattern=gompus+floccosus -> /assets/application.js
10.0.2.2 - - [10/Dec/2020:23:37:41 UTC] "GET /assets/Hygrocybe.css HTTP/1.1" 304 0
http://localhost:3000/observer/observation_search?pattern=gompus+floccosus -> /assets/Hygrocybe.css

```

**Comment by JoeCohen** [comment]
I'm beginning to think I did a typo in the search string. I will have a look at this again tomorrow.

**Comment by JoeCohen** [comment]
I deleted several of my new comments; I now think that the PR behaved correctly and the supposed issues were due to my typos in searches.

**Comment by JoeCohen** [comment]
I'm turning draft PR.

**Comment by nimmolo** [comment]
@pellaea — The test failures on my reboot of this PR also have to do with titles. I don't really understand the title mechanism well enough to figure it out. Is this something you remember enough about to offer a clue?

> Ugh, the failing tests all had to do with query titles. That is definitely a larger problem than I have time to solve completely. (How do we get observation queries to describe themselves in English when there are so many possible permutations of parameters??) But I did take a stab at it, trying at least to preserve the titles for the flavor=all versions of the old flavors like by_user, of_name, at_location, etc. Almost certainly not perfect, but at least all the tests pass again!
> 
> This is definitely something that needs to happen eventually if we ever want to convert all those old "flavorful" queries into "unflavored" (i.e. flavor = :all) queries using filter parameters instead. (Which would be a useful way to make various queries and indexes more transparent and RESTful and persistent. "Observations By User" then becomes a generic observation query with "user=xxx" filter parameter, and we can automatically fill in the search bar with "user:xxx". Then it will be very easy for users to modify the query in nontrivial and useful ways by simply adding terms to the search bar. And so on and on. You get the point, I'm sure. Useful direction to move, I think, just going to be a long journey...

**Comment by nimmolo** [comment]
#1449

**Pull request closed without merging**